File: workers/orchestrator/circuit_breaker.py
"""
Circuit Breaker — pauses a task type after N consecutive failures.

After 5 consecutive failures of the **same task type**, the circuit opens
(pauses) for 60 seconds. After that, it transitions to half-open, allows one
test execution, and either closes (success) or re-opens (failure).

Design
------
    - Consecutive failures are tracked per task type in Redis.
    - States: ``CLOSED`` (normal), ``OPEN`` (paused), ``HALF_OPEN`` (testing).
    - The ``check_circuit()`` function is called before task dispatch.
    - The ``record_failure()`` / ``record_success()`` functions are called
      after task completion.

Redis Keys
----------
    ``stas:circuit:{task_type}:state`` — ``CLOSED`` | ``OPEN`` | ``HALF_OPEN``
    ``stas:circuit:{task_type}:failure_count`` — integer
    ``stas:circuit:{task_type}:opened_at`` — ISO timestamp when circuit opened
    ``stas:circuit:{task_type}:half_open_at`` — ISO timestamp when half-open

Configuration (env vars)
------------------------
    ``CIRCUIT_BREAKER_THRESHOLD`` (default: 5) — consecutive failures before open.
    ``CIRCUIT_BREAKER_OPEN_SECONDS`` (default: 60) — how long circuit stays open.
    ``CIRCUIT_BREAKER_HALF_OPEN_MAX`` (default: 1) — max tests in half-open state.
"""

from __future__ import annotations

import json
import logging
import os
import time
from typing import Any, Optional

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_THRESHOLD = int(os.getenv("CIRCUIT_BREAKER_THRESHOLD", "5"))
_OPEN_SECONDS = int(os.getenv("CIRCUIT_BREAKER_OPEN_SECONDS", "60"))
_HALF_OPEN_MAX = int(os.getenv("CIRCUIT_BREAKER_HALF_OPEN_MAX", "1"))
_REDIS_PREFIX = "stas:circuit:"

# ── States ────────────────────────────────────────────────────────────

CLOSED = "CLOSED"
OPEN = "OPEN"
HALF_OPEN = "HALF_OPEN"

CircuitState = str  # Literal[CLOSED, OPEN, HALF_OPEN]


# ---------------------------------------------------------------------------
# Redis client (lazy)
# ---------------------------------------------------------------------------

_REDIS_CLIENT: Optional[Any] = None


def _get_redis() -> Optional[Any]:
    global _REDIS_CLIENT
    if _REDIS_CLIENT is not None:
        return _REDIS_CLIENT
    try:
        import redis as _redis_mod

        url = os.getenv(
            "REDIS_URL",
            os.getenv("CELERY_RESULT_BACKEND", "redis://localhost:6379/0"),
        )
        _REDIS_CLIENT = _redis_mod.from_url(url, decode_responses=True)
        _REDIS_CLIENT.ping()
        return _REDIS_CLIENT
    except Exception as exc:
        logger.warning("Circuit breaker - Redis unavailable: %s", exc)
        _REDIS_CLIENT = None
        return None


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _state_key(task_type: str) -> str:
    return _REDIS_PREFIX + task_type + ":state"


def _fail_count_key(task_type: str) -> str:
    return _REDIS_PREFIX + task_type + ":failure_count"


def _opened_at_key(task_type: str) -> str:
    return _REDIS_PREFIX + task_type + ":opened_at"


def _half_open_at_key(task_type: str) -> str:
    return _REDIS_PREFIX + task_type + ":half_open_at"


# ---------------------------------------------------------------------------
# State machine
# ---------------------------------------------------------------------------


def get_state(task_type: str) -> CircuitState:
    """Get the current circuit breaker state for a task type.

    Automatically transitions OPEN -> HALF_OPEN after ``_OPEN_SECONDS``.
    """
    client = _get_redis()
    if not client:
        return CLOSED  # Degrade gracefully: closed

    try:
        state = client.get(_state_key(task_type))
        if state is None:
            return CLOSED

        if state == OPEN:
            opened_at = client.get(_opened_at_key(task_type))
            if opened_at:
                try:
                    elapsed = time.time() - float(opened_at)
                    if elapsed >= _OPEN_SECONDS:
                        # Auto-transition to HALF_OPEN
                        set_state(task_type, HALF_OPEN)
                        client.set(_half_open_at_key(task_type), str(time.time()))
                        logger.info(
                            json.dumps({
                                "event": "circuit.auto_half_open",
                                "task_type": task_type,
                                "elapsed_s": round(elapsed, 1),
                            })
                        )
                        return HALF_OPEN
                except (ValueError, TypeError):
                    pass
            return OPEN

        elif state == HALF_OPEN:
            return HALF_OPEN

        return CLOSED
    except Exception as exc:
        logger.debug("Failed to get circuit state for %s - %s", task_type, exc)
        return CLOSED


def set_state(task_type: str, state: CircuitState) -> None:
    """Set the circuit breaker state for a task type."""
    client = _get_redis()
    if not client:
        return

    try:
        client.set(_state_key(task_type), state)
        client.expire(_state_key(task_type), _OPEN_SECONDS * 4)  # safety TTL
    except Exception as exc:
        logger.debug("Failed to set circuit state for %s - %s", task_type, exc)


def check_circuit(task_type: str) -> tuple[bool, str]:
    """Check whether a task is allowed to execute.

    Args:
        task_type: The Celery task name.

    Returns:
        ``(allowed, reason)``. If the circuit is open, ``allowed`` is False
        and ``reason`` explains why.
    """
    state = get_state(task_type)

    if state == OPEN:
        opened_at = _get_opened_at(task_type)
        reason = (
            "Circuit OPEN for task '%s' - paused for %ds (opened at %s)"
            % (task_type, _OPEN_SECONDS, opened_at or "unknown")
        )
        logger.warning(
            json.dumps({
                "event": "circuit.blocked",
                "task_type": task_type,
                "state": state,
                "opened_at": opened_at or "unknown",
            })
        )
        return False, reason

    elif state == HALF_OPEN:
        # In half-open, we allow a limited number of test executions
        test_count = _get_half_open_test_count(task_type)
        if test_count >= _HALF_OPEN_MAX:
            reason = (
                "Circuit HALF_OPEN for task '%s' - already used %d/%d test attempts"
                % (task_type, test_count, _HALF_OPEN_MAX)
            )
            logger.debug("Half-open test limit reached for %s", task_type)
            return False, reason

        _increment_half_open_test_count(task_type)
        logger.info(
            json.dumps({
                "event": "circuit.half_open_test",
                "task_type": task_type,
                "test_count": test_count + 1,
                "max_tests": _HALF_OPEN_MAX,
            })
        )
        return True, "half_open_test"

    return True, "circuit_closed"


def _get_opened_at(task_type: str) -> Optional[str]:
    """Get the ISO timestamp when the circuit was opened."""
    client = _get_redis()
    if not client:
        return None
    try:
        return client.get(_opened_at_key(task_type))
    except Exception:
        return None


# ---------------------------------------------------------------------------
# Recording outcomes
# ---------------------------------------------------------------------------


def record_failure(task_type: str) -> dict[str, Any]:
    """Record a task failure and potentially open the circuit.

    Args:
        task_type: The Celery task name.

    Returns:
        A dict with keys ``state``, ``failure_count``, ``threshold``.
    """
    client = _get_redis()
    if not client:
        return {"state": CLOSED, "failure_count": 0}

    try:
        # Increment consecutive failure count
        fail_count = client.incr(_fail_count_key(task_type))
        client.expire(_fail_count_key(task_type), _OPEN_SECONDS * 4)

        # If we're in HALF_OPEN and get a failure, go back to OPEN
        current_state = get_state(task_type)
        if current_state == HALF_OPEN:
            set_state(task_type, OPEN)
            client.set(_opened_at_key(task_type), str(time.time()))
            logger.warning(
                json.dumps({
                    "event": "circuit.re_opened",
                    "task_type": task_type,
                    "failure_count": fail_count,
                    "threshold": _THRESHOLD,
                })
            )
            return {"state": OPEN, "failure_count": int(fail_count), "threshold": _THRESHOLD}

        # If consecutive failures exceed threshold, open the circuit
        if fail_count >= _THRESHOLD and current_state == CLOSED:
            set_state(task_type, OPEN)
            client.set(_opened_at_key(task_type), str(time.time()))
            logger.warning(
                json.dumps({
                    "event": "circuit.opened",
                    "task_type": task_type,
                    "failure_count": fail_count,
                    "threshold": _THRESHOLD,
                    "pause_duration_s": _OPEN_SECONDS,
                })
            )

        return {
            "state": OPEN if int(fail_count) >= _THRESHOLD else CLOSED,
            "failure_count": int(fail_count),
            "threshold": _THRESHOLD,
        }

    except Exception as exc:
        logger.error("Failed to record failure for %s - %s", task_type, exc)
        return {"state": CLOSED, "failure_count": 0}


def record_success(task_type: str) -> dict[str, Any]:
    """Record a task success and potentially close the circuit.

    Args:
        task_type: The Celery task name.

    Returns:
        A dict with the new state.
    """
    client = _get_redis()
    if not client:
        return {"state": CLOSED}

    try:
        # Reset failure count
        client.delete(_fail_count_key(task_type))

        current_state = get_state(task_type)

        if current_state == HALF_OPEN:
            # Success in half-open means circuit closes
            set_state(task_type, CLOSED)
            client.delete(_opened_at_key(task_type))
            client.delete(_half_open_at_key(task_type))
            _reset_half_open_test_count(task_type)
            logger.info(
                json.dumps({
                    "event": "circuit.closed",
                    "task_type": task_type,
                })
            )
            return {"state": CLOSED}

        elif current_state == OPEN:
            # Success while open is unusual but possible if auto-recovered
            # Keep the circuit open until the timeout expires
            pass

        return {"state": current_state}

    except Exception as exc:
        logger.error("Failed to record success for %s - %s", task_type, exc)
        return {"state": CLOSED}


# ---------------------------------------------------------------------------
# Half-open test tracking (via Redis)
# ---------------------------------------------------------------------------


def _half_open_test_key(task_type: str) -> str:
    return _REDIS_PREFIX + task_type + ":half_open_tests"


def _get_half_open_test_count(task_type: str) -> int:
    client = _get_redis()
    if not client:
        return 0
    try:
        val = client.get(_half_open_test_key(task_type))
        return int(val) if val else 0
    except (ValueError, TypeError, Exception):
        return 0


def _increment_half_open_test_count(task_type: str) -> int:
    client = _get_redis()
    if not client:
        return 0
    try:
        count = client.incr(_half_open_test_key(task_type))
        client.expire(_half_open_test_key(task_type), _OPEN_SECONDS + 10)
        return int(count)
    except Exception:
        return 0


def _reset_half_open_test_count(task_type: str) -> None:
    client = _get_redis()
    if not client:
        return
    try:
        client.delete(_half_open_test_key(task_type))
    except Exception:
        pass


# ---------------------------------------------------------------------------
# Diagnostic / admin
# ---------------------------------------------------------------------------


def get_all_circuits() -> dict[str, dict[str, Any]]:
    """Get the state of all tracked circuit breakers.

    Returns:
        Dict mapping task type -> state info.
    """
    client = _get_redis()
    if not client:
        return {}

    circuits: dict[str, dict[str, Any]] = {}
    try:
        cursor = 0
        while True:
            cursor, keys = client.scan(cursor, match=_REDIS_PREFIX + "*:state")
            for key in keys:
                task_type = key[len(_REDIS_PREFIX):-6]  # strip prefix and ":state"
                state = client.get(key)
                fail_count_raw = client.get(_fail_count_key(task_type))
                opened_at = client.get(_opened_at_key(task_type))
                half_open_at = client.get(_half_open_at_key(task_type))

                info: dict[str, Any] = {
                    "state": state or CLOSED,
                    "failure_count": int(fail_count_raw) if fail_count_raw else 0,
                    "threshold": _THRESHOLD,
                }
                if opened_at:
                    info["opened_at"] = opened_at
                if half_open_at:
                    info["half_open_at"] = half_open_at
                circuits[task_type] = info
            if cursor == 0:
                break
    except Exception as exc:
        logger.error("Failed to list circuits - %s", exc)

    return circuits


def reset_circuit(task_type: str) -> bool:
    """Manually reset a circuit breaker to CLOSED."""
    client = _get_redis()
    if not client:
        return False
    try:
        keys_to_delete = [
            _state_key(task_type),
            _fail_count_key(task_type),
            _opened_at_key(task_type),
            _half_open_at_key(task_type),
            _half_open_test_key(task_type),
        ]
        client.delete(*keys_to_delete)
        logger.info(
            json.dumps({
                "event": "circuit.manual_reset",
                "task_type": task_type,
            })
        )
        return True
    except Exception as exc:
        logger.error("Failed to reset circuit for %s - %s", task_type, exc)
        return False

File: src/monitoring/circuitBreaker.ts
/**
 * Circuit Breaker — pauses a task type after N consecutive failures.
 *
 * Mirrors the Python circuit breaker in workers/orchestrator/circuit_breaker.py
 * for use by TypeScript components (webhook server, queue consumers).
 *
 * States: CLOSED (normal) -> OPEN (paused) -> HALF_OPEN (testing) -> CLOSED/OPEN
 *
 * After 5 consecutive failures of the same task type, the circuit opens
 * (pauses) for 60 seconds. After that, it transitions to half-open, allows one
 * test execution, and either closes (success) or re-opens (failure).
 *
 * Redis Keys:
 *   syntaro:circuit:{taskType}:state         - CLOSED | OPEN | HALF_OPEN
 *   syntaro:circuit:{taskType}:failure_count - integer
 *   syntaro:circuit:{taskType}:opened_at     - Unix timestamp (seconds)
 */

import { Redis } from 'ioredis';
import { config } from '../config.js';
import { rootLogger } from '../utils/logger.js';

const log = rootLogger.child({ module: 'circuit-breaker' });

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const THRESHOLD = Number(process.env.CIRCUIT_BREAKER_THRESHOLD) || 5;
const OPEN_SECONDS = Number(process.env.CIRCUIT_BREAKER_OPEN_SECONDS) || 60;
const HALF_OPEN_MAX = Number(process.env.CIRCUIT_BREAKER_HALF_OPEN_MAX) || 1;
const REDIS_PREFIX = 'syntaro:circuit:';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type CircuitState = 'CLOSED' | 'OPEN' | 'HALF_OPEN';

export interface CircuitInfo {
  state: CircuitState;
  failureCount: number;
  threshold: number;
  openedAt?: number;
  halfOpenAt?: number;
}

// ---------------------------------------------------------------------------
// Redis client (lazy singleton)
// ---------------------------------------------------------------------------

let _redis: Redis | null = null;

function getRedis(): Redis {
  if (!_redis) {
    _redis = new Redis(config.queue.redisUrl, {
      maxRetriesPerRequest: null,
      enableReadyCheck: true,
      retryStrategy: (times: number) => {
        const delay = Math.min(times * 100, 3000);
        log.warn({ attempt: times }, `Redis connection retry in ${delay}ms`);
        return delay;
      },
      lazyConnect: true,
    });
    _redis.on('error', (err) => {
      log.error({ err: String(err) }, 'Circuit breaker Redis error');
    });
  }
  return _redis;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function stateKey(taskType: string): string {
  return `${REDIS_PREFIX}${taskType}:state`;
}

function failCountKey(taskType: string): string {
  return `${REDIS_PREFIX}${taskType}:failure_count`;
}

function openedAtKey(taskType: string): string {
  return `${REDIS_PREFIX}${taskType}:opened_at`;
}

function halfOpenTestKey(taskType: string): string {
  return `${REDIS_PREFIX}${taskType}:half_open_tests`;
}

// ---------------------------------------------------------------------------
// State Management
// ---------------------------------------------------------------------------

/**
 * Get the current circuit breaker state for a task type.
 * Automatically transitions OPEN -> HALF_OPEN after OPEN_SECONDS.
 */
export async function getState(taskType: string): Promise<CircuitState> {
  const redis = getRedis();

  try {
    const state = await redis.get(stateKey(taskType));
    if (!state) return 'CLOSED';

    if (state === 'OPEN') {
      const openedAt = await redis.get(openedAtKey(taskType));
      if (openedAt) {
        const elapsed = Date.now() / 1000 - Number(openedAt);
        if (elapsed >= OPEN_SECONDS) {
          await setState(taskType, 'HALF_OPEN');
          log.info(
            { taskType, elapsedSec: Math.round(elapsed) },
            'Circuit auto-transitioned to HALF_OPEN',
          );
          return 'HALF_OPEN';
        }
      }
      return 'OPEN';
    }

    if (state === 'HALF_OPEN') return 'HALF_OPEN';

    return 'CLOSED';
  } catch (err) {
    log.debug({ err: String(err), taskType }, 'Failed to get circuit state');
    return 'CLOSED';
  }
}

async function setState(taskType: string, state: CircuitState): Promise<void> {
  const redis = getRedis();

  try {
    await redis.set(stateKey(taskType), state, 'EX', OPEN_SECONDS * 4);
  } catch (err) {
    log.debug({ err: String(err), taskType }, 'Failed to set circuit state');
  }
}

/**
 * Check whether a task is allowed to execute.
 * @returns Object with `allowed` and `reason`.
 */
export async function checkCircuit(taskType: string): Promise<{ allowed: boolean; reason: string }> {
  const state = await getState(taskType);

  if (state === 'OPEN') {
    const redis = getRedis();
    let openedAt: string | null = null;
    try {
      openedAt = await redis.get(openedAtKey(taskType));
    } catch (err) {
      log.warn({ err: String(err), taskType }, 'Failed to read circuit breaker opened timestamp');
    }
    const reason = `Circuit OPEN for '${taskType}' - paused for ${OPEN_SECONDS}s (opened at ${openedAt ?? 'unknown'})`;
    log.warn({ taskType, state, openedAt }, 'Circuit breaker blocked task');
    return { allowed: false, reason };
  }

  if (state === 'HALF_OPEN') {
    const redis = getRedis();
    let testCount = 0;
    try {
      testCount = Number(await redis.get(halfOpenTestKey(taskType))) || 0;
      if (testCount >= HALF_OPEN_MAX) {
        const reason = `Circuit HALF_OPEN for '${taskType}' - used ${testCount}/${HALF_OPEN_MAX} test attempts`;
        return { allowed: false, reason };
      }
      await redis.incr(halfOpenTestKey(taskType));
      await redis.expire(halfOpenTestKey(taskType), OPEN_SECONDS + 10);
    } catch (err) {
      log.warn({ err: String(err), taskType }, 'Failed to update half-open test count');
    }
    return { allowed: true, reason: 'half_open_test' };
  }

  return { allowed: true, reason: 'circuit_closed' };
}

/**
 * Record a task failure and potentially open the circuit.
 */
export async function recordFailure(taskType: string): Promise<CircuitInfo> {
  const redis = getRedis();

  try {
    const failCount = await redis.incr(failCountKey(taskType));
    await redis.expire(failCountKey(taskType), OPEN_SECONDS * 4);

    const currentState = await getState(taskType);

    if (currentState === 'HALF_OPEN') {
      await setState(taskType, 'OPEN');
      await redis.set(openedAtKey(taskType), String(Date.now() / 1000));
      log.warn({ taskType, failCount, threshold: THRESHOLD }, 'Circuit re-opened from HALF_OPEN');
      return { state: 'OPEN', failureCount: failCount, threshold: THRESHOLD, openedAt: Date.now() / 1000 };
    }

    if (failCount >= THRESHOLD && currentState === 'CLOSED') {
      await setState(taskType, 'OPEN');
      await redis.set(openedAtKey(taskType), String(Date.now() / 1000));
      log.warn(
        { taskType, failCount, threshold: THRESHOLD, pauseDuration: OPEN_SECONDS },
        'Circuit opened',
      );
      return { state: 'OPEN', failureCount: failCount, threshold: THRESHOLD, openedAt: Date.now() / 1000 };
    }

    return { state: failCount >= THRESHOLD ? 'OPEN' : 'CLOSED', failureCount: failCount, threshold: THRESHOLD };
  } catch (err) {
    log.error({ err: String(err), taskType }, 'Failed to record failure');
    return { state: 'CLOSED', failureCount: 0, threshold: THRESHOLD };
  }
}

/**
 * Record a task success and potentially close the circuit.
 */
export async function recordSuccess(taskType: string): Promise<CircuitInfo> {
  const redis = getRedis();

  try {
    await redis.del(failCountKey(taskType));
    const currentState = await getState(taskType);

    if (currentState === 'HALF_OPEN') {
      await setState(taskType, 'CLOSED');
      await redis.del(openedAtKey(taskType));
      await redis.del(halfOpenTestKey(taskType));
      log.info({ taskType }, 'Circuit closed after successful half-open test');
      return { state: 'CLOSED', failureCount: 0, threshold: THRESHOLD };
    }

    return { state: currentState, failureCount: 0, threshold: THRESHOLD };
  } catch (err) {
    log.error({ err: String(err), taskType }, 'Failed to record success');
    return { state: 'CLOSED', failureCount: 0, threshold: THRESHOLD };
  }
}

/**
 * Get the state of all tracked circuit breakers.
 */
export async function getAllCircuits(): Promise<Record<string, CircuitInfo>> {
  const redis = getRedis();
  const circuits: Record<string, CircuitInfo> = {};

  try {
    const keys = await redis.keys(`${REDIS_PREFIX}*:state`);
    for (const key of keys) {
      const taskType = key.slice(REDIS_PREFIX.length, -6); // strip prefix and ":state"
      const [state, failCountRaw, openedAt] = await Promise.all([
        redis.get(key),
        redis.get(failCountKey(taskType)),
        redis.get(openedAtKey(taskType)),
      ]);
      circuits[taskType] = {
        state: (state as CircuitState) || 'CLOSED',
        failureCount: Number(failCountRaw) || 0,
        threshold: THRESHOLD,
        openedAt: openedAt ? Number(openedAt) : undefined,
      };
    }
  } catch (err) {
    log.error({ err: String(err) }, 'Failed to list circuits');
  }

  return circuits;
}

/**
 * Manually reset a circuit breaker to CLOSED.
 */
export async function resetCircuit(taskType: string): Promise<boolean> {
  const redis = getRedis();

  try {
    await redis.del(stateKey(taskType));
    await redis.del(failCountKey(taskType));
    await redis.del(openedAtKey(taskType));
    await redis.del(halfOpenTestKey(taskType));
    log.info({ taskType }, 'Circuit manually reset');
    return true;
  } catch (err) {
    log.error({ err: String(err), taskType }, 'Failed to reset circuit');
    return false;
  }
}
